File: app_manager/appmanager/src/main/java/app/editors/manager/mvp/models/explorer/CloudFile.java
package app.editors.manager.mvp.models.explorer;

import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;

import java.io.Serializable;

import app.editors.manager.mvp.models.base.Base;
import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.ui.adapters.holders.factory.TypeFactory;

public class CloudFile extends Item implements Serializable, Entity, Cloneable {

    @SerializedName("folderId")
    @Expose
    private String folderId = "";

    @SerializedName("version")
    @Expose
    private int version = 0;

    @SerializedName("versionGroup")
    @Expose
    private String versionGroup = "";

    @SerializedName("contentLength")
    @Expose
    private String contentLength = "";

    @SerializedName("pureContentLength")
    @Expose
    private long pureContentLength = 0;

    @SerializedName("fileStatus")
    @Expose
    private String fileStatus = "";

    @SerializedName("viewUrl")
    @Expose
    private String viewUrl = "";

    @SerializedName("webUrl")
    @Expose
    private String webUrl = "";

    @SerializedName("fileType")
    @Expose
    private String fileType = "";

    @SerializedName("fileExst")
    @Expose
    private String fileExst = "";

    @SerializedName("comment")
    @Expose
    private String comment = "";

    public String getFolderId() {
        return folderId;
    }

    public void setFolderId(String folderId) {
        this.folderId = folderId;
    }

    public int getVersion() {
        return version;
    }

    public void setVersion(int version) {
        this.version = version;
    }

    public String getVersionGroup() {
        return versionGroup;
    }

    public void setVersionGroup(String versionGroup) {
        this.versionGroup = versionGroup;
    }

    public String getContentLength() {
        return contentLength;
    }

    public void setContentLength(String contentLength) {
        this.contentLength = contentLength;
    }

    public long getPureContentLength() {
        return pureContentLength;
    }

    public void setPureContentLength(long pureContentLength) {
        this.pureContentLength = pureContentLength;
    }

    public String getFileStatus() {
        return fileStatus;
    }

    public void setFileStatus(String fileStatus) {
        this.fileStatus = fileStatus;
    }

    public String getViewUrl() {
        return viewUrl;
    }

    public void setViewUrl(String viewUrl) {
        this.viewUrl = viewUrl;
    }

    public String getWebUrl() {
        return webUrl;
    }

    public void setWebUrl(String webUrl) {
        this.webUrl = webUrl;
    }

    public String getFileType() {
        return fileType;
    }

    public void setFileType(String fileType) {
        this.fileType = fileType;
    }

    public String getFileExst() {
        return fileExst;
    }

    public void setFileExst(String fileExst) {
        this.fileExst = fileExst;
    }

    public String getComment() {
        return comment;
    }

    public void setComment(String comment) {
        this.comment = comment;
    }

    public int getNextVersion() {
        return ++version;
    }

    public void setFile(CloudFile file) {
        setItem(file);
        folderId = file.getFolderId();
        version = file.getVersion();
        versionGroup = file.getVersionGroup();
        contentLength = file.getContentLength();
        pureContentLength = file.getPureContentLength();
        fileStatus = file.getFileStatus();
        viewUrl = file.getViewUrl();
        webUrl = file.getWebUrl();
        fileType = file.getFileType();
        fileExst = file.getFileExst();
        comment = file.getComment();
    }

    @Override
    public CloudFile clone() {
        return (CloudFile) super.clone();
    }

    @Override
    public int getType(TypeFactory factory) {
        return factory.type(this);
    }

    /*
    * Comparators
    * */
    public static class SortFilesType extends Base.AbstractSort<CloudFile> {

        public SortFilesType(boolean isSortAsc) {
            super(isSortAsc);
        }

        @Override
        public int compare(CloudFile o1, CloudFile o2) {
            return mSortOrder * o1.getFileExst().compareTo(o2.getFileExst());
        }
    }

    public static class SortFilesSize extends Base.AbstractSort<CloudFile> {

        public SortFilesSize(boolean isSortAsc) {
            super(isSortAsc);
        }

        @Override
        public int compare(CloudFile o1, CloudFile o2) {
            return mSortOrder * o1.getContentLength().compareTo(o2.getContentLength());
        }
    }

}

File: app_manager/core/src/main/java/app/documents/core/network/manager/models/explorer/CloudFile.kt
package app.documents.core.network.manager.models.explorer

import app.documents.core.model.cloud.Access
import app.documents.core.network.common.contracts.ApiContract
import app.documents.core.network.common.models.BaseResponse
import com.google.gson.annotations.Expose
import com.google.gson.annotations.SerializedName
import lib.toolkit.base.managers.utils.StringUtils.Extension
import lib.toolkit.base.managers.utils.StringUtils.getExtension
import java.time.Duration
import java.time.Instant
import java.util.Date

open class CloudFile : Item() {

    @SerializedName("folderId")
    @Expose
    var folderId = ""

    @SerializedName("version")
    @Expose
    var version = 0

    @SerializedName("versionGroup")
    @Expose
    var versionGroup = ""

    @SerializedName("contentLength")
    @Expose
    var contentLength = ""

    @SerializedName("pureContentLength")
    @Expose
    var pureContentLength: Long = 0

    @SerializedName("fileStatus")
    @Expose
    var fileStatus = 0

    @SerializedName("viewUrl")
    @Expose
    var viewUrl = ""

    @SerializedName("webUrl")
    @Expose
    var webUrl = ""

    @SerializedName("fileType")
    @Expose
    var fileType = ""

    @SerializedName("fileExst")
    @Expose
    var fileExst = ""

    @SerializedName("comment")
    @Expose
    var comment = ""

    @SerializedName("locked")
    @Expose
    var isLocked = false

    @SerializedName("formFillingStatus")
    @Expose
    var formFillingStatusType = 0

    @SerializedName("canWebRestrictedEditing")
    @Expose
    var isCanWebRestrictedEditing = true

    @SerializedName("canFillForms")
    @Expose
    var isCanFillForms = true

    @SerializedName("denyDownload")
    @Expose
    var isDenyDownload = false

    @SerializedName("denySharing")
    @Expose
    var isDenySharing = false

    @SerializedName("encrypted")
    @Expose
    var encrypted = false

    @SerializedName("availableExternalRights")
    @Expose
    var availableExternalRights: ExternalAccessRights? = null

    @SerializedName("isForm")
    @Expose
    var isForm: Boolean = false

    @SerializedName("expired")
    val expired: Date? = null

    val nextVersion: Int
        get() = ++version

    val clearExt: String
        get() = fileExst.replace(".", "")

    val isPdfForm: Boolean
        get() = getExtension(fileExst) == Extension.PDF && isForm

    val isFavorite: Boolean
        get() = (fileStatus and ApiContract.FileStatus.FAVORITE) != 0

    val isNew: Boolean
        get() = (fileStatus and ApiContract.FileStatus.IS_NEW) != 0

    val isEditing: Boolean
        get() = (fileStatus and ApiContract.FileStatus.IS_EDITING) != 0

    val isExpiringSoon: Boolean
        get() {
            if (expired == null) return false
            val totalDuration = Duration.between(expired.toInstant(), created.toInstant()).abs()
            val timePassed =  Duration.between(Instant.now(), created.toInstant()).abs()
            return timePassed.toMillis() >= totalDuration.toMillis() * 0.9
        }

    val formFillingStatus: ApiContract.FormFillingStatus
        get() = ApiContract.FormFillingStatus.fromType(formFillingStatusType)

    private fun String.toIntOrZero(): Int {
        return if (isNotEmpty()) toInt() else 0
    }

    override fun clone(): CloudFile {
        return super.clone() as CloudFile
    }

    class SortFilesType(isSortAsc: Boolean) : BaseResponse.AbstractSort<CloudFile>(isSortAsc) {
        override fun compare(o1: CloudFile, o2: CloudFile): Int {
            return mSortOrder * o1.fileExst.compareTo(o2.fileExst)
        }
    }

    class SortFilesSize(isSortAsc: Boolean) : BaseResponse.AbstractSort<CloudFile>(isSortAsc) {
        override fun compare(o1: CloudFile, o2: CloudFile): Int {
            return mSortOrder * o1.contentLength.compareTo(o2.contentLength)
        }
    }
}

val CloudFile.allowShare: Boolean
    get() = access in listOf(Access.Editor, Access.Read) && !isDenySharing